File: vciplreportsv1001/vcipl1001/report/monthwise_sales_report/monthwise_sales_report.js
frappe.query_reports["Monthwise Sales Report"] = {

    filters: [

        {
            fieldname: "customer_group",
            label: __("Customer Group"),
            fieldtype: "Link",
            options: "Customer Group",
            default: "Debtors Distributors"
        },

        {
            fieldname: "customer",
            label: __("Customer"),
            fieldtype: "Link",
            options: "Customer"
        },

        {
            fieldname: "year",
            label: __("Financial Year"),
            fieldtype: "Select",
            options: (function () {

                let years = [];
                let current_year = new Date().getFullYear();

                for (let y = current_year; y >= 2020; y--) {
                    years.push(y.toString());
                }

                return years.join("\n");

            })(),
            reqd: 0
        },

        {
            fieldname: "month",
            label: __("Month"),
            fieldtype: "Select",
            options: [
                "",
                "April","May","June",
                "July","August","September",
                "October","November","December",
                "January","February","March"
            ].join("\n")
        }

    ],


    onload: function(report) {

        let today = new Date();
        let year = today.getFullYear();
        let month = today.getMonth() + 1;

        // Financial Year Logic
        if (month <= 3) {
            year = year - 1;
        }

        report.set_filter_value("year", year.toString());

        setTimeout(() => {
            report.refresh();
        }, 500);
    },


    formatter(value, row, column, data, default_formatter) {

        value = default_formatter(value, row, column, data);

        if (
            column.fieldtype === "Currency" &&
            data &&
            data[column.fieldname + "_drill"] &&
            value > 0 &&
            !data.is_total_row
        ) {

            try {

                let drill_data = JSON.parse(data[column.fieldname + "_drill"]);

                if (drill_data && drill_data.length > 0) {

                    return `<a style="font-weight:bold;color:#1674E0;cursor:pointer;text-decoration:underline;"
                        onclick='frappe.query_reports["Monthwise Sales Report"]
                        .show_drill_down(${JSON.stringify(drill_data)}, "${column.label}", "${data.customer_name}")'>
                        ${value}
                    </a>`;
                }

            } catch(e) {
                return value;
            }
        }

        else if (data && data.is_total_row) {
            return `<span style="font-weight:bold;">${value}</span>`;
        }

        return value;
    },


    show_drill_down(invoices, month, customer) {

        if (!invoices || invoices.length === 0) {
            frappe.msgprint(__("No Sales Invoices for this period"));
            return;
        }

        let total_amount = invoices.reduce((sum, inv) => sum + inv.amount, 0);

        let html = `
        <div style="max-height:500px;overflow:auto">
            <h4 style="margin-bottom:15px; padding-bottom:10px; border-bottom:2px solid #1674E0;">
                ${customer} - ${month}
                <span style="float:right; color:#1674E0; background:#e8f0fe; padding:5px 10px; border-radius:4px;">
                    Total: ${format_currency(total_amount)}
                </span>
            </h4>
            <table class="table table-bordered table-hover">
                <thead style="background-color:#f0f0f0;">
                    <tr>
                        <th style="width:40%;">Sales Invoice</th>
                        <th style="width:30%;">Date</th>
                        <th style="width:30%; text-align:right;">Amount</th>
                    </tr>
                </thead>
                <tbody>`;

        invoices.forEach((inv, index) => {

            html += `
            <tr ${index % 2 === 0 ? 'style="background-color:#fafafa;"' : ''}>
                <td>
                    <a href="/app/sales-invoice/${inv.invoice}"
                       target="_blank"
                       style="font-weight:500; color:#1674E0; text-decoration:none;">
                       ${inv.invoice}
                    </a>
                </td>
                <td>${frappe.datetime.str_to_user(inv.date)}</td>
                <td style="text-align:right; font-weight:500;">${format_currency(inv.amount)}</td>
            </tr>`;
        });

        html += `
                </tbody>
                <tfoot style="background-color:#e8e8e8; font-weight:bold;">
                    <tr>
                        <td colspan="2" style="text-align:right;">Total:</td>
                        <td style="text-align:right; color:#1674E0;">${format_currency(total_amount)}</td>
                    </tr>
                </tfoot>
            </table>
        </div>`;

        frappe.msgprint({
            title: __("Invoice Details: {0} - {1}", [customer, month]),
            message: html,
            wide: true,
            indicator: "green"
        });
    }

};

File: vciplreportsv1001/vcipl1001/report/monthwise_sales_report/monthwise_sales_report.py
import frappe
from frappe.utils import flt
import json
from datetime import datetime


MONTHS = [
    (4, "apr", "April"),
    (5, "may", "May"),
    (6, "jun", "June"),
    (7, "jul", "July"),
    (8, "aug", "August"),
    (9, "sep", "September"),
    (10, "oct", "October"),
    (11, "nov", "November"),
    (12, "dec", "December"),
    (1, "jan", "January"),
    (2, "feb", "February"),
    (3, "mar", "March"),
]


def execute(filters=None):

    filters = filters or {}

    # default financial year
    if not filters.get("year"):

        today = datetime.now()

        if today.month <= 3:
            filters["year"] = str(today.year - 1)
        else:
            filters["year"] = str(today.year)

    columns = get_columns(filters)
    data = get_data(filters)

    return columns, data


def get_columns(filters):

    columns = [{
        "label": "Customer",
        "fieldname": "customer_name",
        "fieldtype": "Data",
        "width": 260
    }]

    selected_month = filters.get("month")

    for month_no, key, label in MONTHS:

        if not selected_month or selected_month == label:

            columns.append({
                "label": label,
                "fieldname": key,
                "fieldtype": "Currency",
                "width": 120
            })

    columns.append({
        "label": "Total",
        "fieldname": "total",
        "fieldtype": "Currency",
        "width": 150
    })

    return columns


def get_data(filters):

    year = filters.get("year")
    customer_group = filters.get("customer_group")
    customer = filters.get("customer")

    from_date = f"{year}-04-01"
    to_date = f"{int(year)+1}-03-31"

    conditions = ["si.docstatus = 1"]

    if customer_group:
        conditions.append("si.customer_group = %(customer_group)s")

    if customer:
        conditions.append("si.customer = %(customer)s")

    where_clause = " AND ".join(conditions)

    invoices = frappe.db.sql(f"""
        SELECT
            si.name AS invoice,
            si.customer_name,
            si.posting_date,
            si.grand_total AS amount,
            MONTH(si.posting_date) AS month_no
        FROM `tabSales Invoice` si
        WHERE {where_clause}
        AND si.posting_date BETWEEN %(from_date)s AND %(to_date)s
        ORDER BY si.customer_name
    """, {
        "from_date": from_date,
        "to_date": to_date,
        "customer_group": customer_group,
        "customer": customer
    }, as_dict=True)

    customer_data = {}
    month_totals = {key: 0 for _, key, _ in MONTHS}

    for inv in invoices:

        cust = inv.customer_name
        month_no = inv.month_no
        amount = flt(inv.amount)

        if cust not in customer_data:

            customer_data[cust] = {
                "customer_name": cust,
                "total": 0
            }

            for m_no, key, label in MONTHS:
                customer_data[cust][key] = 0
                customer_data[cust][f"{key}_drill"] = []

        for m_no, key, label in MONTHS:

            if month_no == m_no:

                customer_data[cust][key] += amount
                customer_data[cust]["total"] += amount

                customer_data[cust][f"{key}_drill"].append({
                    "invoice": inv.invoice,
                    "date": str(inv.posting_date),
                    "amount": amount
                })

                month_totals[key] += amount
                break

    result = []

    for cust, data in customer_data.items():

        row = {
            "customer_name": cust,
            "total": data["total"]
        }

        for m_no, key, label in MONTHS:
            row[key] = data[key]
            row[f"{key}_drill"] = json.dumps(data[f"{key}_drill"])

        result.append(row)

    result.sort(key=lambda x: x["customer_name"])

    if result:

        summary = {
            "customer_name": "<b>GRAND TOTAL</b>",
            "is_total_row": 1,
            "total": sum(x["total"] for x in result)
        }

        for m_no, key, label in MONTHS:
            summary[key] = month_totals[key]
            summary[f"{key}_drill"] = json.dumps([])

        result.append(summary)

    return result